pkg/nextgen/distdoc/graph.go

```diff
@@ -0,0 +1,150 @@
+package distdoc
+
+import (
+	"container/list"
+	"errors"
+	"github.com/nuts-foundation/nuts-network/pkg/model"
+	"sort"
+)
+
+var errRootAlreadyExists = errors.New("root document already exists")
+
+// DAG is a directed acyclic graph consisting of nodes (documents) referring to preceding nodes.
```

Comment:
is there a way to garbage collect this DAG?
does the spec say anything about received documents that are older than MAX_AGE? (not counting actively retrieving old docs)

Reply:
> is there a way to garbage collect this DAG?
Actually yes; the DAG is now rooted by a single document without prev. If the DAG becomes too large I can see 3 options:

- [ ] Introducing a second root so the first one can be GC-ed when the new one contains all relevant information
- [ ] GC-ing just the content of documents that became irrelevant by not including the complete contents in the JWS but just a hash of the contents.
- [ ] Allowing the DAG to be re-rooted, now the root document MUST have no prevs but we could change that, so we can just "forget" about all history starting at document XYZ. Must take care that all relevant information is republished from that document on though.

Any other ideas?

> does the spec say anything about received documents that are older than MAX_AGE? (not counting actively retrieving old docs)

I've thought about it but didn't make it in the end (yet). Most important question is; what potential problem are we trying to solve?

Nothing _should_ break if a node goes rogue by making lots of branches starting at an old node. Although the algorithm to walk the DAG could become resource (memory) hungry because it keeps a queue with current branches to be walked.

As for "age", since we have no central clock age as in clock time isn't of much use (nodes could just spoof their clock time). We could use "older than X documents" but then you could deny other nodes from publishing if you publish many documents at once. Maybe the "older than X" should parameter should be determined by an algorithm?

Reply:
the MAX_AGE comment has correlation to the GC. If everything older than 1 week is GC'ed than accepting documents older than a week requires reloading of the history....

Reply:
Made an issue for it.